libtsuba/src/ParquetReader.cpp

```diff
@@ -122,65 +149,159 @@ tsuba::ParquetReader::ReadFromUriSliced(const katana::Uri& uri) {
+    // TODO(thunt) implement this
     return KATANA_ERROR(
-        tsuba::ErrorCode::ArrowError, "arrow error: {}", open_file_result);
+        ErrorCode::NotImplemented,
+        "sorry! missing support for sliced read when choosing columns");
   }
 
-  std::shared_ptr<arrow::Table> out;
-  auto read_result = reader->ReadTable(&out);
-  if (!read_result.ok()) {
+  if (static_cast<uint64_t>(schema.num_fields()) != filter.size()) {
     return KATANA_ERROR(
-        tsuba::ErrorCode::ArrowError, "arrow error: {}", read_result);
+        ErrorCode::InvalidArgument,
+        "column bitmap length ({}) does not the number of columns ({})",
+        filter.size(), schema.num_fields());
   }
 
+  std::vector<std::shared_ptr<arrow::Field>> fields;
+  std::vector<std::shared_ptr<arrow::ChunkedArray>> columns;
+  for (int64_t i = 0, size = schema.num_fields(); i < size; ++i) {
+    if (!filter[i]) {
+      continue;
+    }
+    std::shared_ptr<arrow::ChunkedArray> column;
+    auto status = reader->ReadColumn(i, &column);
+    if (!status.ok()) {
+      return KATANA_ERROR(
+          ErrorCode::ArrowError, "reading column {}: {}", i, status);
+    }
+    fields.emplace_back(schema.field(i));
+    columns.emplace_back(std::move(column));
+  }
+  return FixTable(arrow::Table::Make(arrow::schema(fields), columns));
+}
+
+Result<std::shared_ptr<arrow::Table>>
+tsuba::ParquetReader::ReadColumn(const katana::Uri& uri, int32_t column_idx) {
+  auto reader_res = MakeFileReader(uri, 0, 0);
+  if (!reader_res) {
+    return reader_res.error();
+  }
+  std::unique_ptr<parquet::arrow::FileReader> reader(
+      std::move(reader_res.value()));
+
+  std::shared_ptr<arrow::Schema> schema;
+  auto status = reader->GetSchema(&schema);
+  if (!status.ok()) {
+    return KATANA_ERROR(ErrorCode::ArrowError, "reading schema: {}", status);
+  }
+
+  if (schema->num_fields() < column_idx) {
+    return KATANA_ERROR(
+        ErrorCode::InvalidArgument,
+        "column_idx ({}) out of bounds for number of columns in {} ({})",
+        column_idx, uri, schema->num_fields());
+  }
+
+  std::vector<bool> filter(schema->num_fields(), false);
+  filter[column_idx] = true;
+
+  return DoFilteredTableRead(reader.get(), *schema, filter);
+}
+
+Result<std::shared_ptr<arrow::Table>>
+tsuba::ParquetReader::ReadTable(
+    const katana::Uri& uri, const std::vector<bool>& column_bitmap) {
+  auto reader_res = MakeFileReader(uri, 0, 0);
+  if (!reader_res) {
+    return reader_res.error();
+  }
+  std::unique_ptr<parquet::arrow::FileReader> reader(
+      std::move(reader_res.value()));
+
+  std::shared_ptr<arrow::Schema> schema;
+  auto status = reader->GetSchema(&schema);
+  if (!status.ok()) {
+    return KATANA_ERROR(ErrorCode::ArrowError, "reading schema: {}", status);
+  }
+
+  return DoFilteredTableRead(reader.get(), *schema, column_bitmap);
+}
+
+Result<int32_t>
+tsuba::ParquetReader::NumColumns(const katana::Uri& uri) {
+  auto reader_res = MakeFileReader(uri, 0, 0);
```

Comment:
Just to check my mental model, I thought we had 1 parquet file per property.  This code is also for import, where who knows what we will find, but for our property tables, it is 1 per file, yes?

Reply:
Ya property files are one column tables. We're pushing people towards giving us data as parquet for import so this is an attempt at making a general set of interfaces for doing that sanely. In particular the ask was to leave some columns out of memory.